### ISSUE TITLE
Stopping criteria for CGSolver

### ISSUE THREAD

**Issue by salomole** [mfem/mfem#3386]
Hello,

I would like to know how the stopping criteria is calculated when I solving a linear system with CGSolver. In fact, I calculate the following error on my unknown variable u :

![image](https://user-images.githubusercontent.com/103731183/209678495-951b80e0-d139-4126-8783-0e722b0f9ace.png)

I impose RTOLERANCE=10^(-12), so I expect to see the error converging to 10^(-6), but it converges to 10^(-4).

**Comment by tzanio** [comment]
Hi @salomole,

If the linear system is $A u = f$ and the CG iterates are $u_0, u_1, \ldots, u_n$ we deine the residual $r_k = f - Au_k$ and measure the relative reduction in its norm:

$$
\frac{||r_n||}{||r_0||} \leq \sqrt{rtol}
$$

i.e.

$$
\sum_i (r_n)_i^2 \leq rtol \sum_i (r_n)_i^2
$$

One can also impose the absolute tolerance condition

$$
\sum_i (r_n)_i^2 \leq atol
$$

but that is really problem-dependent.


Note that $r_n = A e_n$, where $e_n$ is the exact error, $e_n = u_* - u_n$ with $A u_* = f$.

**Comment by salomole** [comment]
Thank you for your answer, but the error E that I calculate (see the definition above) should converge to 10^(-6), right?

**Comment by tzanio** [comment]
Hi @salomole,

Sorry for the delayed response.

If I understand correctly, what you are computing is proportional to the finite element $L^2$ approximation norm $\\| u_{fem} - u_{exact} \\|$. (Because $n \sim h^{-d}$, so $\frac{1}{\sqrt{N}} \sim h^{d/2}$ where $h$ is the mesh size and $d$ is the space dimension.)

Assuming regularity, you can expect that to be $C h^2$ where $C$ is independent of $h$. In particular, it should asymptotically decrease by a factor of 4 under uniform refinement.

This is somewhat orthogonal to the relative tolerance in `CGSolver`. What we are measuring there is the error $\\| u_n - u_{fem} \\|$ (above I denoted $u_{fem}$ with $u_*$)) which measures how well we are solving the finite element linear system as we are iterating.

So to summarize, you can control how close your solution is to $u_{fem}$ by dialing down `RTOLERANCE`, but how close you will be to $u_{exact}$ will depend on the mesh size and the regularity of the problem. You don't usually have direct control over that.

Hope this helps,
Tzanio

**Comment by salomole** [comment]
Hello Tzanio,

Thank you for your answer and your help ! :)
